File: Projects/MultiFileReader/modules/file_handler.py
#
# Nimish Nayak
# 08-06-2017
#

# required libraries
import logging
import os 
from utils import Utils

# File Hander
# This module will handle all the file handling operations 
class FileHandler():

    # initialize the instance variables
    def __init__(self, input_file, output_file=None):
        self.input_file = input_file
        self.output_file = output_file
        logging.debug("File Hander initialized")
 
    # process the input file 
    def process_file(self):

        file_type = Utils.get_file_type(self.input_file)    

        # composition: load the appropriate parser
        # get the tuple of dictionaries for all the users in the data file
        if file_type == ".csv":
            logging.info("parsing the csv file")
            from modules.csv_parser import CsvParser  
            parser = CsvParser(self.input_file)
        elif file_type == ".yml":
            logging.info("parsing the yml file")
            from modules.yml_parser import YmlParser  
            parser = YmlParser(self.input_file)
        elif file_type == ".xml":
            logging.info("parsing the xml file")
            from modules.xml_parser import XmlParser
            parser = XmlParser(self.input_file)
        else:
            # input file is not of the neessary order
            raise IOError("File format accepted are xml,csv and yml") 
        # load the xml file 
        parser.load_file()
        # get the list of dictinary of the xml contents 
        parser.parse_file()
        # process the value 
        parser.process_file()
        # get the processed value from the parser as a string
        self.data = str(parser.get_value())              
    
    # write the output if available else print      
    def write_output(self):

        if not self.output_file:
            logging.info("Writing output to stdout as output file name not provided")
            print self.data   
        else:
            Utils.create_output_directory(self.output_file)
            with open(self.output_file, 'w') as f:
                logging.info("Writing output to the file")
                f.write(self.data)

File: modules/file_handler.php
<?php

#
# Nimish Nayak
# 10-06-2017
#

#
# FileHandler 
#

require_once("./modules/utils.php");

# File Hander
# This module will handle all the file handling operations 

class FileHandler {

    # initialize the instance variables
    function __construct($inputFile, $outputFile) {
        
        $this->inputFile = $inputFile;
        $this->outputFile = $outputFile;
        echo("File Hander initialized\n");

    }

    # process the input file 
    public function processFile() {

        $ext = Utils::getFileExtension($this->inputFile);    

        # composition: load the appropriate parser
        # get the tuple of dictionaries for all the users in the data file
        switch($ext){

            case "csv":
                echo("parsing the csv file\n");
                require_once("./modules/csv_parser.php");
                $parser = new CsvParser($this->inputFile);
                break;

            case "yml":
                echo("parsing the yml file\n");
                require_once("./modules/yml_parser.php");
                $parser = new YmlParser($this->inputFile);
                break;

            case "xml":
                echo("parsing the xml file\n");
                require_once("./modules/xml_parser.php");
                $parser = new XmlParser($this->inputFile);
                break;

            default:
                echo("file type not supported\n");
                return null;
        }

        # read file contents 
        $parser->parseFile();
        # process file contents based on certain criteria 
        $parser->processFile();
        # return parsed value
        $this->data = $parser->getValue();              
    
    }

    # write the output if available else print      
    public function writeOutput() {

        if($this->outputFile != null) {
            Utils::createOutputDirectory($this->outputFile);
            echo("Writing output to the file\n");
            $file = fopen($this->outputFile,"w");
            var_dump($this->data);
            fwrite($file,$this->data);
            fclose($file);
        }
        else
            printf("Writing output to stdout as output file name not provided\n%s\n",$this->data);  
    }

}

?>